sys/trace/trace.c

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2020 Kaspar Schleiser <kaspar@schleiser.de>
+ *                    Freie Universität Berlin
+ *                    Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys_util
+ * @{
+ *
+ * @file
+ * @brief       Execution tracing module implementation
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+
+#include <stdio.h>
+
+#include "irq.h"
+#include "xtimer.h"
+
+#ifndef TRACE_BUFSIZE
+#define TRACE_BUFSIZE 512
+#endif
+
+typedef struct {
+    uint32_t time;
+    uint32_t val;
+} tracebuf_entry_t;
+
+static tracebuf_entry_t tracebuf[TRACE_BUFSIZE];
+static size_t tracebuf_pos;
+
+void trace(uint32_t val)
+{
+    unsigned state = irq_disable();
+    tracebuf[tracebuf_pos % TRACE_BUFSIZE] = (tracebuf_entry_t){ .time = xtimer_now_usec(), .val = val };
+    tracebuf_pos++;
```

Comment:
Should the value be clamped? Is a modulo operation expensive?

Reply:
If so maybe add a mention in the documentation. if someone was tracing in a loop for example it could eventually get overrun depending on the arch.

Reply:
yeah, modulo is on the expensive side, unless there's an fpu. also quite large if it has to import software fp support. what do you mean with 'overrun'?

Reply:
> unless there's an fpu.

... or the value is a power of two.

Reply:
Bad choice of words, I mean that it will eventually start overriding buffer content if not `dumped` or `reset`. Just documenting this in the header should be enough.